File: app/src/main/java/com/solutions/alphil/zambiajobalerts/classes/Job.java
package com.solutions.alphil.zambiajobalerts.classes;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Job {
    @SerializedName("id")
    private int id;

    @SerializedName("title")
    private RenderedTitle title;

    @SerializedName("excerpt")
    private RenderedExcerpt excerpt;

    @SerializedName("content")
    private RenderedContent content;

    @SerializedName("date")
    private String date;

    @SerializedName("link")
    private String link;

    @SerializedName("_embedded")
    private Embedded embedded;

    @SerializedName("meta")
    private Map<String, Object> meta;

    @SerializedName("job-categories")
    private List<Integer> jobCategories;

    @SerializedName("job-types")
    private List<Integer> jobTypes;

    @SerializedName("uagb_excerpt")
    private String uagbExcerpt;

    // Fields for caching support
    private String cachedCompany;
    private String cachedLocation;
    private String cachedJobType;
    private String cachedApplication;

    private static final Map<String, Integer> JOB_TYPE_MAPPING = Map.of(
            "Freelance", 9,
            "Full Time", 6,
            "Internship", 10,
            "Part Time", 7,
            "Temporary", 8,
            "Consultancy", 30,
            "Contract", 31,
            "Consultant", 30,
            "Tender", 32
    );

    public Job() {}

    public static Job fromEntity(JobEntity entity) {
        Job job = new Job();
        job.id = entity.getId();
        job.title = new RenderedTitle(entity.getTitle());
        job.excerpt = new RenderedExcerpt(entity.getExcerpt());
        job.content = new RenderedContent(entity.getContent());
        job.date = entity.getDate();
        job.link = entity.getLink();
        
        Embedded embedded = new Embedded();
        Media media = new Media(entity.getFeaturedImage());
        embedded.featuredMedia = new ArrayList<>();
        embedded.featuredMedia.add(media);
        job.embedded = embedded;

        job.cachedCompany = entity.getCompany();
        job.cachedLocation = entity.getLocation();
        job.cachedJobType = entity.getJobType();
        job.cachedApplication = entity.getApplicationLink();
        
        return job;
    }

    public static class RenderedTitle {
        @SerializedName("rendered")
        private String rendered;
        public RenderedTitle(String rendered) { this.rendered = rendered; }
        public String getRendered() { return rendered; }
    }

    public static class RenderedExcerpt {
        @SerializedName("rendered")
        private String rendered;
        public RenderedExcerpt(String rendered) { this.rendered = rendered; }
        public String getRendered() { return rendered; }
    }

    public static class RenderedContent {
        @SerializedName("rendered")
        private String rendered;
        public RenderedContent(String rendered) { this.rendered = rendered; }
        public String getRendered() { return rendered; }
    }

    public static class Embedded {
        @SerializedName("wp:featuredmedia")
        private List<Media> featuredMedia;

        @SerializedName("author")
        private List<Author> authors;

        @SerializedName("wp:term")
        private List<List<Term>> terms;
    }

    public static class Media {
        @SerializedName("source_url")
        private String sourceUrl;
        public Media(String sourceUrl) { this.sourceUrl = sourceUrl; }
        public String getSourceUrl() { return sourceUrl; }
    }

    public static class Author {
        @SerializedName("name")
        private String name;
        public String getName() { return name; }
    }

    public static class Term {
        @SerializedName("id")
        private int id;
        @SerializedName("name")
        private String name;
        @SerializedName("taxonomy")
        private String taxonomy;
        public int getId() { return id; }
        public String getName() { return name; }
        public String getTaxonomy() { return taxonomy; }
    }

    public int getId() { return id; }
    public String getTitle() { return title != null ? title.getRendered() : ""; }
    public String getContent() { return content != null ? content.getRendered() : ""; }
    public String getExcerpt() {
        if (uagbExcerpt != null && !uagbExcerpt.isEmpty()) return uagbExcerpt;
        return excerpt != null ? excerpt.getRendered() : "";
    }
    public String getDate() { return date; }
    public String getLink() { return link; }
    public String getFeaturedImage() {
        return embedded != null && embedded.featuredMedia != null && !embedded.featuredMedia.isEmpty() ?
                embedded.featuredMedia.get(0).getSourceUrl() : "";
    }

    public String getCompany() {
        if (cachedCompany != null) return cachedCompany;
        if (meta != null && meta.containsKey("_company_name")) return (String) meta.get("_company_name");
        return "";
    }

    public String getLocation() {
        if (cachedLocation != null) return cachedLocation;
        if (meta != null && meta.containsKey("_job_location")) return (String) meta.get("_job_location");
        return "";
    }

    public String getApplication() {
        if (cachedApplication != null) return cachedApplication;
        if (meta != null && meta.containsKey("_application")) return (String) meta.get("_application");
        return "";
    }

    public String getJobType() {
        if (cachedJobType != null) return cachedJobType;
        if (jobTypes != null && !jobTypes.isEmpty()) {
            int jobTypeId = jobTypes.get(0);
            for (Map.Entry<String, Integer> entry : JOB_TYPE_MAPPING.entrySet()) {
                if (entry.getValue() == jobTypeId) return entry.getKey();
            }
            if (embedded != null && embedded.terms != null) {
                for (List<Term> termList : embedded.terms) {
                    for (Term term : termList) {
                        if ("job_listing_type".equals(term.getTaxonomy()) && term.getId() == jobTypeId) return term.getName();
                    }
                }
            }
        }
        return "";
    }

    public String getFormattedDate() {
        if (date != null && date.length() >= 10) return date.substring(0, 10);
        return date;
    }
}

File: app/src/main/java/com/solutions/alphil/zambiajobalerts/classes/Job.kt
package com.solutions.alphil.zambiajobalerts.classes

import com.google.gson.annotations.SerializedName
import com.solutions.alphil.zambiajobalerts.shared.JobTypeMapper

class Job {
    @SerializedName("id")
    private var id: Int = 0

    @SerializedName("title")
    private var title: RenderedTitle? = null

    @SerializedName("excerpt")
    private var excerpt: RenderedExcerpt? = null

    @SerializedName("content")
    private var content: RenderedContent? = null

    @SerializedName("date")
    private var date: String? = null

    @SerializedName("link")
    private var link: String? = null

    @SerializedName("_embedded")
    private var embedded: Embedded? = null

    @SerializedName("meta")
    private var meta: Map<String, Any>? = null

    @SerializedName("job-categories")
    private var jobCategories: List<Int>? = null

    @SerializedName("job-types")
    private var jobTypes: List<Int>? = null

    @SerializedName("uagb_excerpt")
    private var uagbExcerpt: String? = null

    private var cachedCompany: String? = null
    private var cachedLocation: String? = null
    private var cachedJobType: String? = null
    private var cachedApplication: String? = null

    fun getId(): Int = id

    fun getTitle(): String = title?.rendered.orEmpty()

    fun getContent(): String = content?.rendered.orEmpty()

    fun getExcerpt(): String {
        if (!uagbExcerpt.isNullOrEmpty()) return uagbExcerpt.orEmpty()
        return excerpt?.rendered.orEmpty()
    }

    fun getDate(): String? = date

    fun getLink(): String? = link

    fun getFeaturedImage(): String =
        embedded?.featuredMedia?.firstOrNull()?.sourceUrl.orEmpty()

    fun getCompany(): String {
        cachedCompany?.let { return it }
        return meta?.get("_company_name") as? String ?: ""
    }

    fun getLocation(): String {
        cachedLocation?.let { return it }
        return meta?.get("_job_location") as? String ?: ""
    }

    fun getApplication(): String {
        cachedApplication?.let { return it }
        return meta?.get("_application") as? String ?: ""
    }

    fun getJobType(): String {
        cachedJobType?.let { return it }
        val jobTypeId = jobTypes?.firstOrNull() ?: return ""
        val mappedType = JobTypeMapper.nameForId(jobTypeId)
        if (mappedType.isNotEmpty()) return mappedType

        embedded?.terms?.forEach { termList ->
            termList.forEach { term ->
                if (term.taxonomy == "job_listing_type" && term.id == jobTypeId) {
                    return term.name.orEmpty()
                }
            }
        }

        return ""
    }

    fun getFormattedDate(): String? =
        if (date != null && date.orEmpty().length >= 10) date.orEmpty().substring(0, 10) else date

    class RenderedTitle {
        @SerializedName("rendered")
        var rendered: String? = null

        constructor()

        constructor(rendered: String?) {
            this.rendered = rendered
        }
    }

    class RenderedExcerpt {
        @SerializedName("rendered")
        var rendered: String? = null

        constructor()

        constructor(rendered: String?) {
            this.rendered = rendered
        }
    }

    class RenderedContent {
        @SerializedName("rendered")
        var rendered: String? = null

        constructor()

        constructor(rendered: String?) {
            this.rendered = rendered
        }
    }

    class Embedded {
        @SerializedName("wp:featuredmedia")
        var featuredMedia: MutableList<Media>? = null

        @SerializedName("author")
        var authors: List<Author>? = null

        @SerializedName("wp:term")
        var terms: List<List<Term>>? = null
    }

    class Media {
        @SerializedName("source_url")
        var sourceUrl: String? = null

        constructor()

        constructor(sourceUrl: String?) {
            this.sourceUrl = sourceUrl
        }
    }

    class Author {
        @SerializedName("name")
        var name: String? = null
    }

    class Term {
        @SerializedName("id")
        var id: Int = 0

        @SerializedName("name")
        var name: String? = null

        @SerializedName("taxonomy")
        var taxonomy: String? = null
    }

    companion object {
        @JvmStatic
        fun fromEntity(entity: JobEntity): Job {
            val job = Job()
            job.id = entity.id
            job.title = RenderedTitle(entity.title)
            job.excerpt = RenderedExcerpt(entity.excerpt)
            job.content = RenderedContent(entity.content)
            job.date = entity.date
            job.link = entity.link
            job.embedded = Embedded().apply {
                featuredMedia = mutableListOf(Media(entity.featuredImage))
            }
            job.cachedCompany = entity.company
            job.cachedLocation = entity.location
            job.cachedJobType = entity.jobType
            job.cachedApplication = entity.applicationLink
            return job
        }
    }
}
